Get TimeIntegrators working with SCALAR variables.

jwpeterson opened the pull request:
There were a few different issues:
- ODETimeDerivative kernel residuals were being computed even when
  KT_NONTIME was requested.
- The ODETimeDerivative kernel was not assembling into the KT_TIME
  residual the way corresponding TimeKernels do.
- The implicit = false parameter was not respected by ParsedODEKernel,
  so the explicit time integration methods were getting the wrong
  answers.

The fixes for these things were pretty simple:
- Add new intermediate base class, ODETimeKernel.
  This class is responsible for overriding computeResidual() to make
  sure that ODE time derivative Kernels contribute to the proper
  residual, similarly to the way that TimeKernel works.
- Inherit the TransientInterface parameters in ScalarKernel, so that
  'implicit = false' is respected for ScalarKernels.
- Added a test verifying that Crank-Nicolson converges to the correct
  value.  Previously it was converging to a completely wrong value.

Closes #6289.

jwpeterson committed:
Get TimeIntegrators working with SCALAR variables.

There were a few different issues:
* ODETimeDerivative kernel residuals were being computed even when
  KT_NONTIME was requested.
* The ODETimeDerivative kernel was not assembling into the KT_TIME
  residual the way corresponding TimeKernels do.
* The implicit = false parameter was not respected by ParsedODEKernel,
  so the explicit time integration methods were getting the wrong
  answers.

The fixes for these things were pretty simple:
* Add new intermediate base class, ODETimeKernel.
  This class is responsible for overriding computeResidual() to make
  sure that ODE time derivative Kernels contribute to the proper
  residual, similarly to the way that TimeKernel works.
* Inherit the TransientInterface parameters in ScalarKernel, so that
  'implicit = false' is respected for ScalarKernels.
* Added a test verifying that Crank-Nicolson converges to the correct
  value.  Previously it was converging to a completely wrong value.

Closes #6289.

andrsd commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -1253,6 +1254,12 @@ NonlinearSystem::computeResidualInternal(Moose::KernelType type)
+        ODETimeKernel * ode_time = dynamic_cast<ODETimeKernel *>(scalar_kernel.get());
+        if (type == Moose::KT_NONTIME && ode_time)
+          continue;
+
```
Maybe we should consider doing the same deal with 3 warehouses i.e. one warehouse for each `ALL`, `TIME` and `NONTIME` scalar kernels like we do for normal kernels.

This implementation is not a big when people have just a few scalar kernels, but I am not sure what that will do in RELAP-7 where there is going to be possibly hundreds of them.

jwpeterson replied:
OK, I can do this optimization, however this PR fixes the immediate serious bugs with TimeIntegrators for SCALAR variables if you think that would be worth getting merged sooner rather than later?

jwpeterson committed:
Separate ScalarKernels into time/non-time warehouses.

This is an optimization that allows us to avoid checking ScalarKernel
types in the middle of computeResidualInternal().

Closes #6289.

andrsd commented:
👍
